sctest: refactor cumulative tests

postamar opened the pull request:
This commit comprehensively refactors the cumulative test definitions for the declarative schema changer.

Fixes #106920.

Release note: None

Commit:
sctest: refactor cumulative tests

This commit comprehensively refactors the cumulative test definitions
for the declarative schema changer.

Fixes: #106920

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/106743)
<!-- Reviewable:end -->

postamar commented on pkg/ccl/schemachangerccl/BUILD.bazel:
```diff
@@ -93,34 +79,22 @@ sctest_gen(
         "Backup",
+        "BackupMixedVersion",
         "EndToEndSideEffects",
+        "ExecuteWithDMLInjection",
```
`ExecuteWithDMLInjection ` tests were missing in CCL for no good reason.

postamar commented on pkg/ccl/schemachangerccl/BUILD.bazel:
```diff
@@ -93,34 +79,22 @@ sctest_gen(
         "GenerateSchemaChangeCorpus",
         "Pause",
         "Rollback",
+        "ValidateMixedVersionElements",
```
`ValidateMixedVersionElements` tests were missing in CCL for no good reason.

postamar commented on pkg/ccl/schemachangerccl/testdata/end_to_end/create_index/create_index.definition:
```diff
@@ -1,43 +1,43 @@
+INSERT INTO t VALUES ($stageKey + 2, 'c');
+DELETE FROM t WHERE name = 'a' and id = $stageKey;
+INSERT INTO t VALUES ($stageKey, 'a');
+UPDATE t SET name = 'a' WHERE id > 0;
```
This whole test definition needed some love in order to pass the `ExecuteWithDMLInjection` test.

postamar commented on pkg/sql/schemachanger/sctest/test_server_factory.go:
```diff
@@ -0,0 +1,141 @@
+		t *testing.T,
+		fn func(s serverutils.TestServerInterface, tdb *gosql.DB),
+	)
+}
```
`TestServerFactory` and its vanilla and multi-region implementations replace the old `newClusterFunc`.

postamar reviewed (commented):
Review guide: at some point, this PR is going to require a leap of faith from the reviewer. Personally, I'm cool with that because I'm not convinced that the existing tests were that well understood in the first place.

All of the pre-existing tests have been either maintained or fixed. Test cases are no longer randomly skipped.

fqazi reviewed (approved):
Excellent work this is pretty clean! <img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 29 of 33 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106743#-:-N_uEGTvEXX_nuv4L2i3:b-oyhrya)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/sql/schemachanger/sctest/backup.go` line 411 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106743#-N_uCcCvAc9JFTAVXd2i:-N_uCcCvAc9JFTAVXd2j:bguou2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedb6d58f6c699be24022194280b56a7c2f40c90/pkg/sql/schemachanger/sctest/backup.go#L411)):*
> ```Go
> 			// Clean up.
> 			// Hack to deal with corrupt userfiles tables due to #76764.
> 			const vq = `SELECT * FROM crdb_internal.invalid_objects WHERE database_name != 'backups'`
> ```

Can we clean this up now?
___
*[`pkg/ccl/schemachangerccl/testdata/end_to_end/create_index/create_index.definition` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106743#gh-1270207398:-N_uDKCz46c9calEqYmL:b6zrrk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b712157fd8208b8b1848b2d4dd2b5b61a050fb3c/pkg/ccl/schemachangerccl/testdata/end_to_end/create_index/create_index.definition#L11)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This whole test definition needed some love in order to pass the `ExecuteWithDMLInjection` test.
</blockquote></details>

Good catch!


<!-- Sent from Reviewable.io -->

postamar commented:
Thanks for the review! There's still one outstanding issue to resolve in this PR in that the `schemachangerccl` now has an unholy amount of tests. I got rid of the randomized skipping because I didn't like it. I'd much rather we deliberately skip entire classes of checks in regular CI and skip nothing during the nightlies. For instance, in the backup tests, skip everything except RESTORE DATABASE cases for successful schema changes in non-mixed clusters. I'm not sure how to do this properly.

> Can we clean this up now?

I hadn't even looked into this. Presumably! I'll give it a whirl.

postamar commented:
I'm battling the `maybe_stress` portion of the CI. See https://cockroachlabs.slack.com/archives/CJ0H8Q97C/p1690203313338359 for details.

Commit:
github-pull-request-make: temporary workaround

This temporary workaround allows the changes in the previous commit to
pass CI.

Release note: None

fqazi commented:
> @fqazi also I renamed the `TestValidateMixedVersionElements` to `TestPauseMixedVersion` because from what I can tell it's a mixed-version flavour of `TestPause`.
That sounds good, and its a less verbose description.

> I got rid of the randomized skipping because I didn't like it. I'd much rather we deliberately skip entire classes of checks in regular CI and skip nothing during the nightlies.
I think that's a good approach and I'm guessing local builds also run everything too

postamar commented:
bors r+
